keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -443,15 +459,45 @@ class ApplicationService(
     }
   }
 
+  private fun getArtifactVersionByPromotionStatus(
+    environmentName: String,
+    promotionStatus: PromotionStatus,
+    version: String?,
+    artifactInfoInEnvironment: Map<String, List<StatusInfoForArtifactInEnvironment>>,
+    allVersions: List<PublishedArtifact>
+  ): PublishedArtifact? {
+    //only CURRENT and PREVIOUS are supported, as they can be sorted by deploy_at
+    require(promotionStatus in listOf(CURRENT, PREVIOUS)) { "Invalid promotion status used to query" }
+    val ourEnv = artifactInfoInEnvironment[environmentName] ?: return null
+    // we sort the versions by their deployed at time in descending order, then we take
+    // the first one where all the conditions match.
+    val chosenVersion = if (version == null) {
+      ourEnv.sortedByDescending { it.deployedAt }.firstOrNull { it.status == promotionStatus }
+    } else {
+      ourEnv.sortedByDescending { it.deployedAt }.firstOrNull { it.status == promotionStatus && it.replacedByVersion == version}
+    } ?: return null
+    return allVersions.find { it.version == chosenVersion.version }
+  }
+
   // Pinning is a special case when is coming to creating a compare link between versions.
   // If there is a pinned version, which is not the same as the current version, we need
   // to make sure we are creating the comparable link with reference to the pinned version.
```

Comment:
```suggestion
  // Pinning is a special case when it comes to creating a compare link between versions.
  // If there is a pinned version, which is not the same as the current version, we need
  // to make sure we are creating the compare link with reference to the pinned version.
```

Reply:
not changed in my PR, not going to touch.